File: wiki-spark/java/src/main/java/pageRank/Original.java
package pageRank;

import java.util.Arrays;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaHadoopRDD;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import pageRank.dataTypes.WikipediaRevision;
import scala.Tuple2;

public class Original {
	public static void main(String[] args) {
		String inputFile = args[0];
		String outputFile = args[1];
		int iterations = Integer.parseInt(args[2]);
		
		SparkConf conf = new SparkConf().setAppName("Original Page Rank");
		Configuration hadoopConf = new Configuration();
		JavaSparkContext sc = new JavaSparkContext(conf);
		sc.hadoopConfiguration().set("textinputformat.record.delimiter","\n\n");
		
		JavaRDD<WikipediaRevision> revisions = sc.newAPIHadoopFile(inputFile, TextInputFormat.class, 
				LongWritable.class, Text.class,hadoopConf)
		.map(v -> v._2.toString())
		.map(record -> {
			String[] lines = record.split("\n");
			String[] revision = lines[0].split(" ");
			Long revisionId =  Long.parseLong(revision[2]);
			String articleId = revision[3].trim();
			String[] mainArray = lines [3].split(" ", 2);
			String outlinks = null;
			if(mainArray.length == 2) {
				outlinks = mainArray[1];
			}
			return new WikipediaRevision(articleId, revisionId, outlinks);
		});
		
		JavaRDD<WikipediaRevision> singleRevisions = revisions.groupBy(revision -> revision.getArticleId())
		.mapValues(revs -> {
			long latestRevisionId  = 0;
	        String articleId = null;
	        String outlinks = null;
	        for(WikipediaRevision revision : revs)
	        {
		          if(revision.getRevisionId() > latestRevisionId)
		          {
		            articleId = revision.getArticleId();
		            latestRevisionId = revision.getRevisionId();
		            outlinks = revision.getOutlinks();
		          }
	        }
	        return new WikipediaRevision(articleId, latestRevisionId, outlinks);
		}).map(t -> t._2);
		/*
		singleRevisions.mapToPair(r -> {
			
		});
		
		JavaPairRDD<String, Iterable<String>> links = lines
				.mapToPair( s ->
				String[] parts = s.split('\\s+');
				return new Tuple2<String, String>(parts[0], parts[1]); )
				.distinct().groupByKey().cache();
				*/
	}
}

File: wiki-spark/scala/src/main/scala/pageRank/Original.scala
package pageRank

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.Text

import pageRank.dataTypes.WikipediaRevision
import pageRank.dataTypes.WikipediaRevision
import org.apache.spark.api.java.JavaPairRDD
import java.util.ArrayList

object Original {
  def main(args: Array[String]) {
    val inputFile = args(0)
    val outputFile = args(1)
    val iterations = args(2).toInt

    val conf = new SparkConf().setAppName("Original Page Rank")
    val sc = new SparkContext(conf)
    sc.hadoopConfiguration.set("textinputformat.record.delimiter", "\n\n")

    val revisions = sc.newAPIHadoopFile(inputFile, classOf[TextInputFormat], classOf[LongWritable], classOf[Text])
      .map(v => v._2.toString())
      .map(record => {
        val lines = record.split("\n")
        val revision = lines(0).split(" ")
        val revisionId = revision(2).trim().toLong
        val articleId = revision(3).trim()
        val mainArray = lines(3).split(" ", 2)
        var outlinks: String = null
        if (mainArray.length == 2) {
          outlinks = mainArray(1)
        }
        new WikipediaRevision(articleId, revisionId, outlinks)
      })

    val singleRevisions = revisions.groupBy(revision => revision.articleId)
      .mapValues(revisions => {
        var latestRevisionId: Long = 0
        var articleId = ""
        var outlinks = ""

        for (revision <- revisions) {
          if (revision.revisionId > latestRevisionId) {
            articleId = revision.articleId
            latestRevisionId = revision.revisionId
            outlinks = revision.outlinks
          }
        }
        new WikipediaRevision(articleId, latestRevisionId, outlinks)
      })
      .map(t => t._2)

    val links = singleRevisions
      .keyBy(_.articleId)
      .mapValues(rev => rev.outlinks.split("\\s"))

    var ranks = links.keyBy(_._1).map(rev => (rev._1, 1.0))

    for (i <- 1 to iterations) {
      val contribs = links.join(ranks).values
        .flatMap((v) => {
          var res: List[Tuple2[String, Double]] = List[Tuple2[String, Double]]()
          var urlCount = v._1.length
          for (s <- v._1) {
            val tuple = new Tuple2[String, Double](s, v._2 / urlCount)
            res = res.+:(tuple)
          }
          res
        })
      ranks = contribs.reduceByKey((a, b) => a + b).mapValues(v => 0.15 + v * 0.85);
    }

    ranks.saveAsTextFile(outputFile)

  }
}